crates/router/src/core/payment_methods.rs

```diff
@@ -565,3 +565,49 @@ pub async fn retrieve_payment_method_with_token(
     };
     Ok(token)
 }
+
+#[cfg(all(feature = "v2", feature = "payment_methods_v2"))]
+pub struct PaymentMethodAddServer;
+#[cfg(all(feature = "v2", feature = "payment_methods_v2"))]
+pub struct PaymentMethodAddClient;
+
+#[cfg(all(feature = "v2", feature = "payment_methods_v2"))]
+#[async_trait::async_trait]
+pub trait PaymentMethodAdd<T: Send + Sync> {
+    async fn perform_preprocessing(
+        &self,
+        state: &SessionState,
+        req: &api::PaymentMethodCreate,
+        merchant_account: &domain::MerchantAccount,
+        key_store: &domain::MerchantKeyStore,
+        data: T,
+    ) -> RouterResult<T>;
+    async fn vault_payment_method(
+        &self,
+        state: &SessionState,
+        req: &api::PaymentMethodCreate,
+        merchant_account: &domain::MerchantAccount,
+        key_store: &domain::MerchantKeyStore,
+        data: T,
+    ) -> RouterResult<T>;
+    async fn handle_duplication(
+        &self,
+        state: &SessionState,
+        req: &api::PaymentMethodCreate,
+        merchant_account: &domain::MerchantAccount,
+        key_store: &domain::MerchantKeyStore,
+        data: T,
+    ) -> RouterResult<T>;
+}
+
+#[cfg(all(feature = "v2", feature = "payment_methods_v2"))]
+pub struct PaymentMethodVaultingData {
+    pub pm_id: Option<String>,
+    pub payment_method: Option<diesel_models::PaymentMethod>,
```

Comment:
We don't have a equivalent domain type for this?

Reply:
Not yet, will implement domain model for Payment Method along with the diesel related changes.